### PR TITLE
laserfix and cone implementation

### PR DESCRIPTION
**Pull request by kacperoo29**

### CODE REVIEW

**Comment by danil179** [comment]
Open new PR or amend commit details so you coukd sign the CLA.

Also did you checked that the laser name string don't break any old spell? (I saw yoi chaned lucian Q and lux R you sure this is all laser spells?)

Cone seems good to me, check if Acos is the fastest math function for cos, but we coukd change it later.

I saw you changed projectile spell data to protected, it's ok to me (so laser could apply later, dont see better way)

But, please open/check existing issue according to the guidelines in the wiki, and add here Progres/Resolve in PR about what issue this PR solve.

Also check if this PR doesnt break style guidelines in the readme.

File: GameServerLib/Logic/GameObjects/Missiles/Laser.cs [hunk 1]
```diff
@@ -27,11 +27,13 @@ internal class Laser : Projectile
             AttackableUnit owner,
             Target target,
             Spell originSpell,
+            string nameLaser,
```
Comment (MythicManiac) [inline GameServerLib/Logic/GameObjects/Missiles/Laser.cs]: `effectName` would be more descriptive

File: GameServerLib/Logic/GameObjects/Spells/Spell.cs [hunk 2]
```diff
@@ -255,7 +255,7 @@ public void AddProjectileTarget(string nameMissile, Target target, bool isServer
             }
         }
 
-        public void AddLaser(float toX, float toY, bool affectAsCastIsOver = true)
+        public void AddLaser(string nameLaser, float toX, float toY, bool affectAsCastIsOver = true)
```
Comment (MythicManiac) [inline GameServerLib/Logic/GameObjects/Spells/Spell.cs]: `effectName` would be more descriptive

File: GameServerLib/Logic/GameObjects/Spells/Spell.cs [hunk 3]
```diff
@@ -265,12 +265,31 @@ public void AddLaser(float toX, float toY, bool affectAsCastIsOver = true)
...
                 new Target(toX, toY),
                 this,
+                nameLaser,
                 SpellData.Flags,
                 affectAsCastIsOver
             );
             _game.ObjectManager.AddObject(l);
         }
 
+        public void AddCone(string nameCone, float toX, float toY, float angleDeg, bool affectAsCastIsOver = true)
```
Comment (MythicManiac) [inline GameServerLib/Logic/GameObjects/Spells/Spell.cs]: `effectName` would be more descriptive

File: GameServerLib/Logic/GameObjects/Missiles/Cone.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+        public Cone(
+            Game game,
+            float x,
+            float y,
+            int collisionRadius,
+            AttackableUnit owner,
+            Target target,
+            Spell originSpell,
+            string nameCone,
```
Comment (MythicManiac) [inline GameServerLib/Logic/GameObjects/Missiles/Cone.cs]: `effectName` would be more descriptive

**Review by MythicManiac — COMMENTED** [review]
A good contribution, however there are some things I'd like to see changed before this is merged. Nice work so far 👍

File: GameServerLib/Logic/GameObjects/Missiles/Cone.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            if (!_affectAsCastIsOver)
+            {
+                return;
+            }
+
+            if (_originSpell.State != SpellState.STATE_CASTING)
+            {
+                var objects = _game.ObjectManager.GetObjects().Values;
+                foreach (var obj in objects)
+                {
```
Comment (MythicManiac) [inline GameServerLib/Logic/GameObjects/Missiles/Cone.cs]: This whole loop should be optimized in a few ways, I'll elaborate below.

1. Split the `CheckFalgsForUnit` to two different functions, one which applies the effects and one which checks if an unit is a valid target. Call the, `CheckIfValidTarget` and `ApplyEffects` for example.
2. In the loop, check unit validity _before_ measuring if said unit is within the hitbox. This saves performance, as checking flags is generally faster than doing the math required for collision check.

**Review by MatthewFrench — APPROVED** [review]

**Comment by MythicManiac** [comment]
This needs to have the merge conflict resolved before it's possible to merge, also you should configure your git identity to your commits.

Let me know on Discord if you need any assistance, I'll be glad to help (and I'm sure others are too)

**Pull request closed without merging**